designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,372 @@
+* 👍 Foundation for Future Extensions: Offers a framework for incorporating other generic disruption controls into the per-reason controls, extending beyond budgets.
+* 👍👍 Customization for Specific Reasons: Allows for specific validations and extensions for individual reasons, offering tailored control.
+* 👎👎👎 Breaking API Changes:Implementing this approach would necessitate significant changes to the current API, especially for budgets. If we decide to model DisruptAfter in a similar way we also would have to break those apis. It might make sense to break the budgets now before they have garnered large adoption as it becomes harder to make the change as time goes on.
+* 👎 Complexity in Default Reason Handling: The model does not facilitate easy defaulting for unspecified ("default") disruption reasons.
+* 👎 Increased Budget Complexity: The use of budgets becomes more complex as they are now nested within another field, adding a layer of intricacy to their application.
+
+### API Design Conclusion: Preferred Design
+After evaluating different approaches to extend the Karpenter API for specifying disruption reasons, the preferred design is the List Approach in Approach A. This approach offers flexibility in managing multiple disruption reasons under a single budget and reduces configuration complexity. It extends the existing API without introducing breaking changes and simplifies management for scenarios where multiple disruption reasons share similar constraints.
+
+While the idea of per-reason controls (Approach B) provides granular control and a foundation for future extensions, it involves significant API changes and increased complexity, making it less favorable at this stage. However, this approach remains a viable option for future considerations, especially if there is a need for more tailored control over each disruption reason.
+
+### Disruption Behavior
+The calculation of allowed disruptions in a system with multiple disruption reasons has become more intricate following the introduction of Disruption Budget Reasons. Previously, the formula was straightforward:
+
+AllowedDisruptions = minNodeCount - unhealthyNodes - totalDisruptingNodes.
+
+When calculating by reason, three potential equations emerge:
+1. BucketedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes[reason]
+2. AllowedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes
+3. minimumDisruptionAllowed = min(minNodeCount[reason], minNodeCount[default]) - unhealthyNodes - totalDisruptingNodes
+
+### Scenarios
+We will go through scenarios that users are expecting to face to drive home if the api can satisfy the users requirements for the end users of these features
+- Block Drift, but allow Expiration
+- Fully Block Drift && Expiration, but allow consolidation to occur 
+- Block Drift && Expiration, but allow emptiness 
+- Limit Drift && Expirationn while allowing agressive but not unbounded consolidationn
```

Comment:
typo: `consolidation`

Reply:
for that note seems Expiration is not spelled as Expirationn